tekton-catalog/pipeline-loops/cmd/cli/main.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+func validateFlags(action, inputFileName, inputFileType string) {
+	var errs []error
+	if action != "validate" {
+		err := fmt.Errorf("unsupported action %s\n", action)
+		errs = append(errs, err)
+	}
+	if inputFileName == "" {
+		err := fmt.Errorf("missing input spec, please specify -f /path/to/file.(yaml|json)\n")
+		errs = append(errs, err)
+	}
+	if inputFileType != "yaml" && inputFileType != "json" {
+		err := fmt.Errorf("unsupported input file type:%s\n", inputFileType)
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		fmt.Println("Errors while processing input:")
+		for i, e := range errs {
+			fmt.Printf("Error: %d -> %v\n", i, e)
+		}
+		os.Exit(1)
+	}
+}
+
+func main() {
+	var action, inputFileName, inputFileType string
+
+	flag.StringVar(&action, "a", "validate", "The `action` on the resource.")
+	flag.StringVar(&inputFileName, "f", "", "path to input `filename` either a yaml or json.")
+	flag.StringVar(&inputFileType, "file-type", "yaml", "`yaml` or a json")
+	flag.Parse()
+	validateFlags(action, inputFileName, inputFileType)
+	if inputFileType == "json" {
+		fmt.Println("Error: json file type is not supported.")
+		os.Exit(5)
+	}
+	errs := []string{}
+	if inputFileType == "yaml" {
+		objs, err := readFile(inputFileName)
+		if err != nil {
+			fmt.Printf("Error while reading input spec: %v\n", err)
+			os.Exit(2)
+		}
+		for _, o := range objs {
+			marshalledBytes, err := o.MarshalJSON()
+			if err != nil {
+				fmt.Printf("Error while marshalling json: %v\n", err)
+				os.Exit(3)
+			}
+			err = nil
+			switch kind := o.GetKind(); kind {
+			case "Task":
+				// No validation.
+			case "Run":
+				err = validateRun(marshalledBytes)
+			case "Pipeline":
+				err = validatePipeline(marshalledBytes)
+			case "PipelineLoop":
+				err = validatePipelineLoop(marshalledBytes)
+			case "PipelineRun":
+				err = validatePipelineRun(marshalledBytes)
+			default:
+				fmt.Printf("Warn: Unsupported kind: %s.\n", kind)
+			}
+			if err != nil {
+				errs = append(errs, err.Error())
+			}
+		}
+		if len(errs) > 0 {
+			fmt.Printf("Validation errors: %s\n", strings.Join(errs, "\n"))
+			os.Exit(100)
+		} else {
+			fmt.Printf("\nCongratulations, all checks passed !!\n")
+		}
+	}
+}
+
+func validatePipeline(bytes []byte) error {
+	p := v1beta1.Pipeline{}
+	err := json.Unmarshal(bytes, &p)
+	if err != nil {
+		return err
+	}
+	return validatePipelineSpec(&p.Spec, p.Name)
+}
+
+func validatePipelineSpec(p *v1beta1.PipelineSpec, name string) error {
+	errs := []string{}
+	// We do not need to validate PipelineRun because it is validated by tekton admission webhook
+	// And pipelineTask.TaskRef is also validated by tekton.
+	// Here we only need to validate those embedded spec, whose kind is pipelineLoop.
+	if p.Tasks != nil {
```

Comment:
you can use tekton validate apis to to it `github.com/tektoncd/pipeline/pkg/apis/validate`